notes/conds.md

```diff
@@ -0,0 +1,84 @@
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+    ["var", ["?x", "?y"], // The `?` here is just by convention
+      ["args", "foo..bar.[].baz", "?x"],
+      [["some", "match"], "?x", "*@example.com"],
+      
+      ["args", "foo.quux", "?y"],
+      [">", "?y", 42],
+      ["<", "?x", "?y"]
+    ]
+  ]
+}
+```
+
+Of note:
+
+* Selector syntax based on JSONPath / JSON Poiniter / jQ
+  * `"foo..bar.[].baz"` ~ `foo.{*recursive descent*}.bar.{ALL}.baz`
+* `var` is like `fresh` in minikanren: introduces logic variables
+  * these give the ability to run predicate logic in the middle of a sector
+  * `[[<node>, <path>, "?x"], [<pred>, "?x", <expr>]`
+* Declarative matching from `args` onwards
+  * Anything of the form `[<node>, <selector>, <value or variable>]`
+  * Can avoid namespace conflicts because users define variable names locally
+  * Recurses if needed
+    * A trivial case: `[["args", "some.path", "?x"], ["?x", "more.path", 42]]`
+      * Matches `args.some.path.more.path == 42`
+  * Use of `args` to start the path allows us to open up the synatx over time if we want
+    * Plus it's very declarative
```

Comment:
cc @Gozala thoughts?

Reply:
Perhaps worth noting that the above also lets us avoid `==` in a lot of common cases, too:
`["args", "some.path", 42]` is like `{==: {args: {some: {path: 42}}}}`

Reply:
Odd thing about `["args", "some.path", "?x"]` is that it's bit unclear if e.g. this would be valid

`["", "args.some.path", "?x"]` or this `["args.some", "path", "?x"]`

I feel like it may make more sense to think of `args` as a built-in pre-poulated variable. perhaps `?` could be that. Or maybe for built-in variables we could use different prefix so we could introduce new ones in the future without colliding with user defined variables.

Reply:
> ["", "args.some.path", "?x"]

I think that we can't run the selector on `""`, so it should fail, right?

>  perhaps ? could be that

Yeah, exatcly: I was thinking of `args` as that pre-populated variable. How about `?args` for consistent variable syntax?

Reply:
> so we could introduce new ones in the future without colliding with user defined variables.

Not a bad idea, though I think it's up to the user to avoid collisions. We have a schema type in the capsule, so we know what's reserved.

We could introduce more syntax like `$env` (`$args`), which feels maybe appropriate? It's _is_ from the $environment